### ISSUE TITLE
How to wait for an on_next operation on subject?

### ISSUE THREAD

**Issue by tonyabracadabra** [ReactiveX/RxPY#347]
I have a subject `subj = Subject()` whose `on_next(event)` will be called asynchronouly on a different thread through NewThreadScheduler

```
observable.subscribe(lambda event: subj.on_next(event), scheduler=NewThreadScheduler())
```

Now I want to listen to the `on_next()` operation so I will know that there is a new event being added to `subj`, how am I suppose to do it?

I am now writing as,

```
subj = ReplaySubject()
scheduler = NewThreadScheduler()

# this is an Observable created from grpc stream
grpc_stream: Observable = client.subscribe_to_workspace(subscription_input)

def subscribe_func(e):
    subj.on_next(e)
    print('completed on_next', subj, e)

grpc_dispose: Disposable = grpc_stream.subscribe(subscribe_func, scheduler=scheduler)

# creating an event to grpc_stream (not immediately), after a while I can see screen print out 'completed on_next' as written the subscribe_func 
send_event()

async def a():
    print('waiting...', subj)

    # it blocks here and never get resolved
    return await subj.pipe(op.take(1))

asyncio.run(a())
```

So I can wait for the first event being added to `subj` until executing other tasks that depend on the creation of the first event through `send_event()`.

The code I tried above doesn't really work as I got blocked at `await subj.pipe(op.take(1)) `, I'm pretty sure that `grpc_stream` has received events, and the `on_next(event)` was executed perfectly with `NewThreadScheduler` so there won't be a block when I am executing the code. However, the line `await subj.pipe(op.take(1))` seems to wait forever and never get solved. Where did I do wrong? Should I use a different scheduler or anything that helps?

**Comment by MainRo** [comment]
Your code should work, however if on_next is called from another thread than the event loop, then it must either be scheduled on the event loop with an AsyncioScheduler (created with threadsafe=True), either be scheduled manually e.g. by using call_soon_threadsafe(subj.on_next, event).

**Comment by tonyabracadabra** [comment]
> Your code should work, however if on_next is called from another thread than the event loop, then it must either be scheduled on the event loop with an AsyncioScheduler (created with threadsafe=True), either be scheduled manually e.g. by using call_soon_threadsafe(subj.on_next, event).

Hi Romain, thanks for your reply. I made that `on_next` operation through ```observable.subscribe(lambda e: subj.on_next(e), scheduler=NewThreadScheduler())``` and it was blocked there. When I switched to your solution by using `observable.subscribe(lambda e: subj.on_next(e), scheduler=AsyncIOScheduler())` it gives me an 
```RuntimeError: There is no current event loop in thread 'MainThread'. ``` exception. This issue has blocked me for days, do you know any way to solve this problem? I sincerely appreciate it.

**Comment by tonyabracadabra** [comment]
> Your code should work, however if on_next is called from another thread than the event loop, then it must either be scheduled on the event loop with an AsyncioScheduler (created with threadsafe=True), either be scheduled manually e.g. by using call_soon_threadsafe(subj.on_next, event).

Hey Romain I've updated the question and hopefully, that clears my current situation out. Thank you so much!

**Comment by MainRo** [comment]
It seems that the call to subscribe is done outside of the event loop. So in this case, the asyncioscheduler should be created before, in the context of the event loop (and with threadsafe=True). So you need to create the scheduler and provide it's reference up to the code that subscribes.

**Comment by MainRo** [comment]
Please do not create several issues for the same question. You can add new comments or edit your previous ones to add explainations if needed.

**Comment by tonyabracadabra** [comment]
> It seems that the call to subscribe is done outside of the event loop. So in this case, the asyncioscheduler should be created before, in the context of the event loop (and with threadsafe=True). So you need to create the scheduler and provide it's reference up to the code that subscribes.

Thanks. Sorry I closed the new issue. I tried using the `AsyncIoScheduler` with,
```
loop = asyncio.new_event_loop()
scheduler = AsyncIOScheduler(loop=loop, threadsafe=True)
```
and when I ran 
```
workspace_dispose: Disposable = workspace_subs.subscribe(func, scheduler=scheduler)
```
it got blocked. 

before I was using `NewThreadScheduler` and that works fine and I can receive events from the stream.

What is there difference and should I use `AsyncIOScheduler` instead of `NewThreadScheduler` in this case? Because seems right now when I use `AsyncIOScheduler`, it is blocked and I cannot receive any event from the observable sequence.

**Comment by tonyabracadabra** [comment]
> Please do not create several issues for the same question. You can add new comments or edit your previous ones to add explainations if needed.

I've updated the description with my current code. Thank you so much!

**Comment by jcafhe** [comment]
> it gives me an
RuntimeError: There is no current event loop in thread 'MainThread'. 

I'm not familliar with asyncio but I think you need to setup an asyncio loop.

It's hard to answer your question because we don't know the context. Are you sure you need a subject? Do you realy need 2 'sub' streams coming from the source (grpc), one for executing code after the 1st item, and a 2nd to do something with each items? 

If you simply need to do something after the 1st item:
```python
import rx
from rx import operators as ops
from rx.concurrency import NewThreadScheduler

new_thread_scheduler = NewThreadScheduler()

# emulates data generated by your grpc request, pushing a incremental
# sequence of int every 0.2s from an other thread
observable_stream = rx.interval(0.2, scheduler=new_thread_scheduler).pipe(
    ops.do_action(lambda i: print('pushing {}'.format(i))),  # log pushed items
    )

# build an observable that completes after 1 item and
# block the current thread until 'first' observable complete (i.e. 1 item)
first = observable_stream.pipe(
    ops.take(1),
    )

first.run()
print('other functions to be executed...')
```

If you want multiple streams, it is more complicated because if you mutate some states after the 1st item is pushed and these states are required to be initialized before beeing used by your second stream, you could have a race condition. 
In that case, you want to defer the subcription to the source until all the downstream subcribers are registered, possibly by using `publish` & `auto_connect`.

**Comment by tonyabracadabra** [comment]
> > it gives me an
> > RuntimeError: There is no current event loop in thread 'MainThread'.
> 
> I'm not familliar with asyncio but I think you need to setup an asyncio loop.
> 
> It's hard to answer your question because we don't know the context. Are you sure you need a subject? Do you realy need 2 'sub' streams coming from the source (grpc), one for executing code after the 1st item, and a 2nd to do something with each items?
> 
> If you simply need to do something after the 1st item:
> 
> ```python
> import rx
> from rx import operators as ops
> from rx.concurrency import NewThreadScheduler
> 
> new_thread_scheduler = NewThreadScheduler()
> 
> # emulates data generated by your grpc request, pushing a incremental
> # sequence of int every 0.2s from an other thread
> observable_stream = rx.interval(0.2, scheduler=new_thread_scheduler).pipe(
>     ops.do_action(lambda i: print('pushing {}'.format(i))),  # log pushed items
>     )
> 
> # build an observable that completes after 1 item and
> # block the current thread until 'first' observable complete (i.e. 1 item)
> first = observable_stream.pipe(
>     ops.take(1),
>     )
> 
> first.run()
> print('other functions to be executed...')
> ```
> 
> If you want multiple streams, it is more complicated because if you mutate some states after the 1st item is pushed and these states are required to be initialized before beeing used by your second stream, you could have a race condition.
> In that case, you want to defer the subcription to the source until all the downstream subcribers are registered, possibly by using `publish` & `auto_connect`.
> 
> EDIT:
> Alternatively, you can work with one stream and use the `do_action` operator to initialize stuff after the 1st element, avoiding the need of multiple subscriptions synchronization.

@jcafhe Hi thank you so much for providing the alternative solution! Since I am receiving events of different kinds, I am actually listening to different subjects with these different events (I am using one subj here for simplification, and in my actual code I will have several `await subj_i.pipe(op.take(1)))` in different places. 

So here is the context, I have a grpc stream that will receive events continuously of different kinds, and I want to make an `observable` from the grpc stream, so I can update `subjects` with the name of the event with `on_next()` (for each event, there is a corresponding `subj`, thus a corresponding line of `await subj.pipe(op.take(1))`), so in the line `await`, the `subj` will receive a signal and complete waiting once the `on_next()` is called.

Also could you elaborate on 
> if you mutate some states after the 1st item is pushed and these states are required to be initialized before beeing used by your second stream
I am not so sure what does the `state` mean in your context.

**Comment by MainRo** [comment]
Here is an example that should be similar to what you want to do:

```
import threading
import time
import rx
from rx.subjects import Subject
import asyncio
from rx import operators as ops
from rx.concurrency import NewThreadScheduler
from rx.concurrency.mainloopscheduler import AsyncIOScheduler

new_thread_scheduler = NewThreadScheduler()

loop = asyncio.new_event_loop()
asyncio.set_event_loop(loop)
aio_scheduler = AsyncIOScheduler(loop=loop,threadsafe=True)

subj = Subject()


def grpc_request(observer, scheduler):
    print(scheduler)
    print("id: {}".format(threading.get_ident()))
    time.sleep(0.5)
    observer.on_next(1)
    observer.on_completed()
    print("request done")


grpc_stream = rx.create(grpc_request)


grpc_dispose = grpc_stream.pipe(
    ops.subscribe_on(new_thread_scheduler),
    ops.observe_on(aio_scheduler),    
).subscribe(
    on_next=subj.on_next, 
    on_completed=subj.on_completed,
    # scheduler=new_thread_scheduler)
)


async def a():
    print("id a: {}".format(threading.get_ident()))
    print('waiting...', subj)

    return await subj.pipe(
        ops.take(1),
    )


print("id main: {}".format(threading.get_ident()))

loop.run_until_complete(a())
print('done')
```

I think that the subscribe_on operator should not be required, but the default scheduler is not propagated up to the grpc_request subscription function.

**Comment by MainRo** [comment]
An update here. With the fix for #356 the subscription can be re-written this way:

```
def grpc_request(observer, scheduler):
    print(scheduler)
    print("id: {}".format(threading.get_ident()))
    def scheduled_request(action, state):
        time.sleep(0.5) # simulate blocking grpc request
        observer.on_next(1)
        observer.on_completed()
    scheduler.schedule(scheduled_request)
    print("request done")

grpc_stream = rx.create(grpc_request)

grpc_dispose = grpc_stream.pipe(
    ops.observe_on(aio_scheduler),    
).subscribe(
    on_next=subj.on_next, 
    on_completed=subj.on_completed,
    scheduler=new_thread_scheduler,
)
```

**Comment by tonyabracadabra** [comment]
> An update here. With the fix for #356 the subscription can be re-written this way:
> 
> ```
> def grpc_request(observer, scheduler):
>     print(scheduler)
>     print("id: {}".format(threading.get_ident()))
>     def scheduled_request(action, state):
>         time.sleep(0.5) # simulate blocking grpc request
>         observer.on_next(1)
>         observer.on_completed()
>     scheduler.schedule(scheduled_request)
>     print("request done")
> 
> grpc_stream = rx.create(grpc_request)
> 
> grpc_dispose = grpc_stream.pipe(
>     ops.observe_on(aio_scheduler),    
> ).subscribe(
>     on_next=subj.on_next, 
>     on_completed=subj.on_completed,
>     scheduler=new_thread_scheduler,
> )
> ```

Thank you Romain! I think it turns out to be a problem in grpc not rxpy, I figured some other way out. Thanks for the code snippet, I learned a lot from it.
